base/array.jl

```diff
@@ -1216,6 +1218,154 @@ function insert!(a::Array{T,1}, i::Integer, item) where T
+    inserteach!(a::Vector, indices::Vector, items::Vector)
+
+Insert all of the elements of `items` into `a` for the given `range` or `indices`. `range` or `indices` are the resulting indices of elements of `items` in the `a`.
+
+# Examples
+```jldoctest
+julia> inserteach!([1, 2, 3], 2:3, [8, 9])
+5-element Array{Int64,1}:
+ 1
+ 8
+ 9
+ 2
+ 3
+```
+
+Following example inserts `10, 9, and 8` at index `6 ,4, and 2` respectively
+```jldoctest
+julia> inserteach!([1, 2, 3, 4, 5], 6:-2:2, [10, 9, 8])
+8-element Array{Int64,1}:
+  1
+  8
+  2
+  9
+  3
+ 10
+  4
+  5
+```
+
+It is possible to use boolean indices (each true shows insertion index):
+```jldoctest
+julia> inserteach!([1, 2, 3], [true, false, true, false, false], [10, 9])
+5-element Array{Int64,1}:
+ 10
+  1
+  9
+  2
+  3
+```
+"""
+function inserteach!(a::Array{T,1}, indices::AbstractVector, items::Array{T,1}) where {T}
+    itemslen = length(items)
+    indiceslen = length(indices)
+    if itemslen !== indiceslen
+        throw(DimensionMismatch("Length of the indices and items should be the same"))
+    elseif itemslen === 0
+        return a
+    end
+    if issorted(indices) # based on the tests overhead/advantage is low
+        indices_sorted = indices
+        items_sorted = items
+    else
+        sort_indices = sortperm(indices) # sorts so the final index is correct
+        indices_sorted = indices[sort_indices]
+        items_sorted = items[sort_indices]
+    end
+    _growat!.(Ref(a), indices_sorted, 1) # does bound check
+    @inbounds for (index, item) in zip(indices_sorted, items_sorted)
+        a[index] = item
+    end
+    return a
+end
+
+
+function inserteach!(a::Array{T,1}, indices::AbstractRange{<:Integer}, items::Array{T,1}) where {T}
+    itemslen = length(items)
+    indiceslen = length(indices)
+    if itemslen !== indiceslen
+        throw(DimensionMismatch("Length of the indices and items should be the same"))
+    elseif itemslen === 0
+        return a
+    end
+    if issorted(indices) # based on the tests overhead/advantage is low
+        indices_sorted = indices
+        items_sorted = items
+    else
+        indices_sorted = sort(indices) # sorting them is faster than indexing into them
+        sort_indices = sortperm(indices) # sorts so the final index is correct
+        items_sorted = items[sort_indices]
+    end
+    _growat!.(Ref(a), indices_sorted, 1) # does bound check
+    @inbounds for (index, item) in zip(indices_sorted, items_sorted)
+        a[index] = item
+    end
+    return a
+end
+
+function inserteach!(a::Array{T,1}, r::AbstractUnitRange{<:Integer}, items::Array{T,1}) where T
+    ilen = length(items)
+    rlen = length(r)
+    if ilen !== rlen
+        throw(DimensionMismatch("Length of the range and items should be the same"))
+    else
+        return inserteach!(a, first(r), items)
+    end
+end
+
+function inserteach!(a::Array{T,1}, boolindices::AbstractVector{Bool}, items::Array{T,1}) where T
```

Comment:
Not sure we need this method. Seems like the user could call `inserteach!(a, findall(boolindices), items)` if they really need this.

Reply:
It is more optimized since it doesn't check for the sort (no `issorted`).